_source/_posts/2021-09-24-java-records.md

```diff
@@ -0,0 +1,556 @@
+
+{% img blog/java-records/openjdk-logo.png alt:"OpenJDK Logo" width:"300" %}{: .center-image }
+
+
+**Prerequisites**:
+
+- [HTTPie](https://httpie.io/)
+- [Java 14+](https://openjdk.java.net/install/index.html)
+- [Okta CLI](https://cli.okta.com)
+
+**Table of Contents**{: .hide }
+* Table of Contents
+{:toc}
+
+# The Record Keyword
+
+The `record` is a new type of declaration, a restricted form of class that acts as a transparent carrier for immutable data. So let's start the exploration by defining a simple data type `EndOfGame` as a `record`:
+
+```java
+import java.time.LocalDate;
+import java.time.LocalTime;
+
+public record EndOfGame(String id, LocalDate date,  LocalTime timeOfDay, String mentalState,
+                        Integer damageTaken, Integer damageToPlayers, Integer damageToStructures) {
+}
+```
+As you can see in the example above, the `record` has a name, `EndOfGame` in the example above. What looks like a constructor signature is the _state description_ or _header_ declaring the _components_ of the `record`.
+
+The following members are acquired automatically with the declaration:
+
+- A private final field and a public read accessor for each component of the state description
+- A public _canonical_ constructor with the same signature as the state description, which initializes each field from the corresponding argument
+- Implementation of `equals()`, `hashCode()` and `toString()`
+
+The code fragments from `EndOfGameTest` class below verify the automatic members are indeed available.
+
+```java
+private EndOfGame getTestEndOfGame(){
+
+    return new EndOfGame("1", LocalDate.of(2018, 12, 12),
+            LocalTime.of(15, 15), 1,
+            "sober", "10%", 1,1,
+            10, 10,10);
+}
+```
+The automatic canonical constructor is used for creating a sample `EndOfGame`.
+
+```java
+@Test
+public void equalsTest(){
+    EndOfGame eog1 = getTestEndOfGame();
+    EndOfGame eog2 = getTestEndOfGame();
+
+    assertTrue(eog1.equals(eog2));
+    assertEquals(eog1, eog2);
+    assertEquals(eog1.hashCode(), eog2.hashCode());
+}
+```
+
+In the test above, `eog1` has the same state than `eog2`, so `eog1.equals(eog2)` is `true`. This also implies both instances have the same `hashCode`.
+
+A `toString()` test can illustrate what the string representation looks like:
+
+```text
+EndOfGame[id=1, date=2018-12-12, timeOfDay=15:15, mentalState=sober, damageTaken=10, damageToPlayers=10, damageToStructures=10]
+```
+
+Automatic read accessors have the same name and return type as the component. Note there is no `get*` prefix in the read accessor name, the same name as the component:
+
+```java
+assertEquals("sober", eog.mentalState());
+```
+
+
+# Java Record Restrictions and Rules
+
+
```

Comment:
It would be nice to have a sentence or two between headers. Can you provide an intro to the section here?